lib/mobius.ex

```diff
@@ -133,6 +133,137 @@ defmodule Mobius do
+  @type csv_opt() ::
+          {:file, String.t()}
+          | {:naming, [naming_opt]}
+
+  @doc """
+  Produces a CSV of currently collected metrics, optionally writing the CSV to file.
+  The CSV looks like: timestamp, name, type, value, tag1, tag2, tag3..., tagN
+  If tags are provided, only the metrics matching the tags are output to the CSV.
+  If a writable file is given via the :file option, the CSV is written to it. Otherwise it is written to the terminal.
+  If the optional :naming option contains :csv_ext, the .csv extension is added to the file name if not already present.
+  If the optional :naming option list :timestamp, the file name is prefixed by a timestamp .
+
+  Examples:
+
+  iex> Mobius.to_csv("vm.memory.total", %{})
+  # -- writes CSV values to the terminal
+
+  iex> Mobius.to_csv("vm.memory.total", %{}, file: "/data/csv/vm.memory.total")
+  # -- writes CSV values to file vm.memory.total
+
+  iex> Mobius.to_csv("vm.memory.total", %{}, file: "/data/csv/vm.memory.total", naming: [:csv_ext, :timestamp])
+  # -- writes CSV values to a file like 20210830T174954_vm.memory.total.csv
+  """
+  @spec to_csv(String.t(), map, [csv_opt]) :: :ok
+  def to_csv(metric_name, tags \\ %{}, opts \\ []) do
+    parsed_metric_name = parse_metric_name(metric_name)
+
+    rows =
+      opts
+      |> Keyword.get(:name, :mobius)
+      |> Scraper.all()
+      |> Enum.flat_map(fn {timestamp, metrics} ->
+        rows_from_metrics(metrics, parsed_metric_name, tags, timestamp)
+      end)
+
+    tag_names = unique_tag_names(rows)
+
+    headers_row =
+      ["timestamp", "name", "type", "value"] ++
+        for tag_name <- tag_names, do: Atom.to_string(tag_name)
+
+    data_rows = data_rows(rows, metric_name, tag_names)
+
+    csv([headers_row | data_rows], opts)
+  end
+
+  defp unique_tag_names(rows) do
+    Enum.reduce(rows, MapSet.new(), fn row, set ->
+      Enum.reduce(Map.keys(row.tags), set, fn tag_name, acc -> MapSet.put(acc, tag_name) end)
+    end)
+    |> Enum.sort()
+  end
+
+  defp data_rows(rows, metric_name, tag_names) do
+    Enum.reduce(rows, [], fn row, acc ->
+      tag_values = for tag_name <- tag_names, do: "#{Map.get(row.tags, tag_name, "")}"
+
+      data_row =
+        ["#{row.timestamp}", "#{metric_name}", "#{row.type}", "#{row.value}"] ++ tag_values
+
+      acc ++ [data_row]
+    end)
+  end
+
+  defp csv(all_rows, opts) do
+    lines = for row <- all_rows, do: Enum.join(row, ",")
+
+    filepath = csv_file(opts)
+
+    out =
+      if filepath do
+        with :ok <- File.mkdir_p(Path.dirname(filepath)),
+             {:ok, device} <- File.open(filepath, [:write]) do
+          device
+        else
+          {:error, reason} ->
+            IO.puts(
+              "Failed to open file #{inspect(filepath)} to write CSV: #{inspect(reason)}. Writing to terminal instead."
+            )
+
+            :stdio
+        end
+      else
+        :stdio
+      end
+
+    for line <- lines, do: IO.write(out, line <> "\n")
```

Comment:
How about delete that `lines = for row <- all_rows, do: Enum.join(row, ",")` and change this to:

```suggestion
  Enum.each(all_rows, fn row -> IO.write(out, [Enum.intersperse(row, ","), "\n"]) end)
```

I'm not sure how that reads to you, but it uses iodata to avoid a whole lot of string concats and removes a big list creation.

Reply:
Done